fix(components): [table] fix the issue where the expand-row-keys property does not support the number type

SpanManX opened the pull request:
修复 expand-row-keys 属性 number 类型无效问题

> Fix the issue where the expand-row-keys property does not support the number type

fix #20203

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

Fix the issue where the expand-row-keys property does not support the number type

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

Fix the issue where the expand-row-keys property does not support the number type

warmthsea commented on packages/components/table/src/store/tree.ts:
```diff
@@ -83,14 +83,15 @@ function useTree<T>(watcherData: WatcherPropsData<T>) {
         if (ifChangeExpandRowKeys) {
           if (expandRowKeys.value) {
-            return ifExpandAll || expandRowKeys.value.includes(key)
+            return ifExpandAll || expandRowKeys.value.some((val) => key == val)
```
The current changes are problematic [play demo](https://element-plus.run/#eyJBcHAudnVlIjoiPHRlbXBsYXRlPlxuICA8ZGl2PlxuICAgIDxlbC10YWJsZVxuICAgICAgOmRhdGE9XCJ0YWJsZURhdGFcIlxuICAgICAgc3R5bGU9XCJ3aWR0aDogMTAwJTsgbWFyZ2luLWJvdHRvbTogMjBweFwiXG4gICAgICByb3cta2V5PVwiaWRcIlxuICAgICAgYm9yZGVyXG4gICAgICA6ZXhwYW5kLXJvdy1rZXlzPVwiZXhwYW5kS2V5c1wiXG4gICAgPlxuICAgICAgPGVsLXRhYmxlLWNvbHVtbiBwcm9wPVwiZGF0ZVwiIGxhYmVsPVwiRGF0ZVwiIHNvcnRhYmxlIC8+XG4gICAgICA8ZWwtdGFibGUtY29sdW1uIHByb3A9XCJuYW1lXCIgbGFiZWw9XCJOYW1lXCIgc29ydGFibGUgLz5cbiAgICAgIDxlbC10YWJsZS1jb2x1bW4gcHJvcD1cImFkZHJlc3NcIiBsYWJlbD1cIkFkZHJlc3NcIiBzb3J0YWJsZSAvPlxuICAgIDwvZWwtdGFibGU+XG4gIDwvZGl2PlxuPC90ZW1wbGF0ZT5cbjxzY3JpcHQgbGFuZz1cInRzXCIgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUnXG5cbmludGVyZmFjZSBVc2VyIHtcbiAgaWQ6IG51bWJlclxuICBkYXRlOiBzdHJpbmdcbiAgbmFtZTogc3RyaW5nXG4gIGFkZHJlc3M6IHN0cmluZ1xuICBoYXNDaGlsZHJlbj86IGJvb2xlYW5cbiAgY2hpbGRyZW4/OiBVc2VyW11cbn1cblxuY29uc3QgZXhwYW5kS2V5cyA9IHJlZihbM10pXG5cbmNvbnN0IHRhYmxlRGF0YTogVXNlcltdID0gW1xuICB7XG4gICAgaWQ6ICczJyxcbiAgICBkYXRlOiAnMjAxNi0wNS0wMScsXG4gICAgbmFtZTogJ3dhbmd4aWFvaHUnLFxuICAgIGFkZHJlc3M6ICdOby4gMTg5LCBHcm92ZSBTdCwgTG9zIEFuZ2VsZXMnLFxuICAgIGNoaWxkcmVuOiBbXG4gICAgICB7XG4gICAgICAgIGlkOiAzMSxcbiAgICAgICAgZGF0ZTogJzIwMTYtMDUtMDEnLFxuICAgICAgICBuYW1lOiAnd2FuZ3hpYW9odScsXG4gICAgICAgIGFkZHJlc3M6ICdOby4gMTg5LCBHcm92ZSBTdCwgTG9zIEFuZ2VsZXMnLFxuICAgICAgfSxcbiAgICAgIHtcbiAgICAgICAgaWQ6IDMyLFxuICAgICAgICBkYXRlOiAnMjAxNi0wNS0wMScsXG4gICAgICAgIG5hbWU6ICd3YW5neGlhb2h1JyxcbiAgICAgICAgYWRkcmVzczogJ05vLiAxODksIEdyb3ZlIFN0LCBMb3MgQW5nZWxlcycsXG4gICAgICB9LFxuICAgIF0sXG4gIH0sXG5dXG48L3NjcmlwdD5cbiIsImVsZW1lbnQtcGx1cy5qcyI6ImltcG9ydCBFbGVtZW50UGx1cyBmcm9tICdlbGVtZW50LXBsdXMnXG5pbXBvcnQgeyBnZXRDdXJyZW50SW5zdGFuY2UgfSBmcm9tICd2dWUnXG5cbmxldCBpbnN0YWxsZWQgPSBmYWxzZVxuYXdhaXQgbG9hZFN0eWxlKClcblxuZXhwb3J0IGZ1bmN0aW9uIHNldHVwRWxlbWVudFBsdXMoKSB7XG4gIGlmIChpbnN0YWxsZWQpIHJldHVyblxuICBjb25zdCBpbnN0YW5jZSA9IGdldEN1cnJlbnRJbnN0YW5jZSgpXG4gIGluc3RhbmNlLmFwcENvbnRleHQuYXBwLnVzZShFbGVtZW50UGx1cylcbiAgaW5zdGFsbGVkID0gdHJ1ZVxufVxuXG5leHBvcnQgZnVuY3Rpb24gbG9hZFN0eWxlKCkge1xuICBjb25zdCBzdHlsZXMgPSBbJ2h0dHBzOi8vcHJldmlldy0yMDIxMC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguY3NzJywgJ2h0dHBzOi8vcHJldmlldy0yMDIxMC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL3RoZW1lLWNoYWxrL2RhcmsvY3NzLXZhcnMuY3NzJ10ubWFwKChzdHlsZSkgPT4ge1xuICAgIHJldHVybiBuZXcgUHJvbWlzZSgocmVzb2x2ZSwgcmVqZWN0KSA9PiB7XG4gICAgICBjb25zdCBsaW5rID0gZG9jdW1lbnQuY3JlYXRlRWxlbWVudCgnbGluaycpXG4gICAgICBsaW5rLnJlbCA9ICdzdHlsZXNoZWV0J1xuICAgICAgbGluay5ocmVmID0gc3R5bGVcbiAgICAgIGxpbmsuYWRkRXZlbnRMaXN0ZW5lcignbG9hZCcsIHJlc29sdmUpXG4gICAgICBsaW5rLmFkZEV2ZW50TGlzdGVuZXIoJ2Vycm9yJywgcmVqZWN0KVxuICAgICAgZG9jdW1lbnQuYm9keS5hcHBlbmQobGluaylcbiAgICB9KVxuICB9KVxuICByZXR1cm4gUHJvbWlzZS5hbGxTZXR0bGVkKHN0eWxlcylcbn1cbiIsInRzY29uZmlnLmpzb24iOiJ7XG4gIFwiY29tcGlsZXJPcHRpb25zXCI6IHtcbiAgICBcInRhcmdldFwiOiBcIkVTTmV4dFwiLFxuICAgIFwianN4XCI6IFwicHJlc2VydmVcIixcbiAgICBcIm1vZHVsZVwiOiBcIkVTTmV4dFwiLFxuICAgIFwibW9kdWxlUmVzb2x1dGlvblwiOiBcIkJ1bmRsZXJcIixcbiAgICBcInR5cGVzXCI6IFtcImVsZW1lbnQtcGx1cy9nbG9iYWwuZC50c1wiXSxcbiAgICBcImFsbG93SW1wb3J0aW5nVHNFeHRlbnNpb25zXCI6IHRydWUsXG4gICAgXCJhbGxvd0pzXCI6IHRydWUsXG4gICAgXCJjaGVja0pzXCI6IHRydWVcbiAgfSxcbiAgXCJ2dWVDb21waWxlck9wdGlvbnNcIjoge1xuICAgIFwidGFyZ2V0XCI6IDMuM1xuICB9XG59XG4iLCJQbGF5Z3JvdW5kTWFpbi52dWUiOiI8c2NyaXB0IHNldHVwPlxuaW1wb3J0IEFwcCBmcm9tICcuL0FwcC52dWUnXG5pbXBvcnQgeyBzZXR1cEVsZW1lbnRQbHVzIH0gZnJvbSAnLi9lbGVtZW50LXBsdXMuanMnXG5zZXR1cEVsZW1lbnRQbHVzKClcbjwvc2NyaXB0PlxuXG48dGVtcGxhdGU+XG4gIDxBcHAgLz5cbjwvdGVtcGxhdGU+XG4iLCJpbXBvcnQtbWFwLmpzb24iOiJ7XG4gIFwiaW1wb3J0c1wiOiB7XG4gICAgXCJ2dWVcIjogXCJodHRwczovL2Zhc3RseS5qc2RlbGl2ci5uZXQvbnBtL0B2dWUvcnVudGltZS1kb21AbGF0ZXN0L2Rpc3QvcnVudGltZS1kb20uZXNtLWJyb3dzZXIuanNcIixcbiAgICBcIkB2dWUvc2hhcmVkXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AdnVlL3NoYXJlZEBsYXRlc3QvZGlzdC9zaGFyZWQuZXNtLWJ1bmRsZXIuanNcIixcbiAgICBcImVsZW1lbnQtcGx1c1wiOiBcImh0dHBzOi8vcHJldmlldy0yMDIxMC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguZnVsbC5taW4ubWpzXCIsXG4gICAgXCJlbGVtZW50LXBsdXMvXCI6IFwidW5zdXBwb3J0ZWRcIixcbiAgICBcIkBlbGVtZW50LXBsdXMvaWNvbnMtdnVlXCI6IFwiaHR0cHM6Ly9mYXN0bHkuanNkZWxpdnIubmV0L25wbS9AZWxlbWVudC1wbHVzL2ljb25zLXZ1ZUAyL2Rpc3QvaW5kZXgubWluLmpzXCJcbiAgfSxcbiAgXCJzY29wZXNcIjoge31cbn0iLCJfbyI6eyJzaG93SGlkZGVuIjp0cnVlLCJzdHlsZVNvdXJjZSI6Imh0dHBzOi8vcHJldmlldy0yMDIxMC1lbGVtZW50LXBsdXMuc3VyZ2Uuc2gvYnVuZGxlL2Rpc3QvaW5kZXguY3NzIn19)

SpanManX replied:
I see the demo is working fine. Is it an issue with the "==" type comparison?

SpanManX replied:
![屏幕截图 2025-03-23 014223](https://github.com/user-attachments/assets/baf40c31-d426-48ed-89aa-bc08b5888a7a)

SpanManX replied:
Use `Map` instead of `Object` to solve the problem of `number` keys

SpanManX committed:
Merge branch 'dev' into fix-expand-row-keys

SpanManX committed:
fix(components): [table] Use Map instead of Object to handle number keys

 Use Map instead of Object to solve the problem of number keys

warmthsea reviewed (commented):
1. `expandRowKeys` should be of type `Array<string | number>`
2. It would be better to add types to `new Map()`
3. It would be better to add relevant test cases

SpanManX committed:
Merge branch 'dev' into fix-expand-row-keys

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

1.expandRowKeys should be of type Array<string | number> ✔
2.It would be better to add types to new Map() ✔
3.It would be better to add relevant test cases ✔

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

1. expandRowKeys should be of type Array<string | number> ✔
2. It would be better to add types to new Map() ✔
3. It would be better to add relevant test cases ✔

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

1. expandRowKeys should be of type Array<string | number> ✔
2. It would be better to add types to new Map() ✔
3. It would be better to add relevant test cases ✔

SpanManX commented:
> 1. `expandRowKeys` should be of type `Array<string | number>`
> 2. It would be better to add types to `new Map()`
> 3. It would be better to add relevant test cases

**Hello, I have modified the code as requested. Please help review it.**

1. expandRowKeys should be of type Array<string | number> ✔
2. It would be better to add types to new Map() ✔
3. It would be better to add relevant test cases ✔

warmthsea commented:
https://github.com/stillgg/element-plus/blob/fix-el-select-v2-model-value/docs/en-US/component/table.md?plain=1#L285

The documentation also needs to be changed, adding:

```md
Starting from version ^(2.9.8), keys of type number are supported.
```

```
Array<string | number>
```

warmthsea commented on packages/components/table/__tests__/table.test.ts:
It would be better to add a test for mixed strings and numbers

SpanManX replied:
1. table.md ✔
2. add a test for mixed strings and numbers ✔

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

1.table.md ✔
2.add a test for mixed strings and numbers ✔

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

1.table.md ✔
2.add a test for mixed strings and numbers ✔

SpanManX commented on packages/components/table/src/store/tree.ts:
```diff
@@ -44,23 +44,23 @@ function useTree<T>(watcherData: WatcherPropsData<T>) {
       data,
       (parent, children, level) => {
-        const parentId = getRowIdentity(parent, rowKey)
+        const parentId = parent[rowKey]
```
> 在 table tree 中 getRowIdentity(parent, rowKey) 函数只执行到
> 
> if (!rowKey.includes('.')) {
>       return \``${row[rowKey]}`\`
>     }
> 
> return \``${row[rowKey]}`\`  会返回 string 类型，并且 getRowIdentity 函数有多个文件在使用，考虑到如果修改 getRowIdentity  函数会影响到其他文件的使用。所以改为 parent[rowKey] 把影响降低到最小，并且 `parent[rowKey]` 不会改变原有数据的类型，获取 `rowKey` 的逻辑也与 getRowIdentity 函数的一致，

In the table tree, the function getRowIdentity(parent, rowKey) only executes up to:

if (!rowKey.includes('.')) {
  return \``${row[rowKey]}`\`
}

return \``${row[rowKey]}`\`  This will return a string type. Since the getRowIdentity function is used in multiple files, modifying it directly may impact other files.

To minimize the impact, it has been changed to parent[rowKey], which does not alter the original data type. Additionally, the logic for retrieving rowKey remains consistent with the getRowIdentity function.

![screenshot-20250327-174309](https://github.com/user-attachments/assets/59c7166e-88e2-4432-8be4-788b2304a1cc)

SpanManX committed:
Merge branch 'dev' into fix-expand-row-keys

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

1. table-test-common.ts ✔
2. table.md ✔
3. tree.ts -- changed to: new Map<string | number, object>() ✔

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

fix(components): [table] fix the bug of the expand-row-keys property

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

fix(components): [table] fix the bug of the expand-row-keys property

SpanManX commented on packages/components/table/src/store/tree.ts:
```diff
@@ -44,23 +44,23 @@ function useTree<T>(watcherData: WatcherPropsData<T>) {
   const normalize = (data) => {
     const rowKey = watcherData.rowKey.value
-    const res = {}
+    const res = new Map<Array<string | number>>() // 使用 Map 替代 Object，解决 number key 的问题
```
changed to:  `const res = new Map<string | number, object>()`

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

fix(components): [table] fix the bug of the expand-row-keys property

SpanManX committed:
fix(components): [table] fix packages/components/table/src/store/tree.ts

fix(components): [table] fix packages/components/table/src/store/tree.ts

warmthsea reviewed (approved):
LGTM, Thanks!

warmthsea committed:
chore: update test cases

SpanManX committed:
Merge branch 'dev' into fix-expand-row-keys

warmthsea committed:
chore: update version

Dsaquel commented on packages/components/table/src/util.ts:
```diff
@@ -177,12 +177,13 @@ export const getColumnByCell = function <T>(
   if (isString(rowKey)) {
     if (!rowKey.includes('.')) {
-      return `${row[rowKey]}`
+      return isReturnRawValue ? row[rowKey] : `${row[rowKey]}`
```
I think it fix for this issue but it should always return `row[rowKey]` whenever the situation.

Many test fails because there are many place like [getKeysMap](https://github.com/SpanManX/element-plus/blob/fix-expand-row-keys/packages/components/table/src/util.ts#L206-L209) that should be a `Map` too like you did in `tree.ts`.

That said it is not blocking this pr.

SpanManX replied:
After changing to row[rowKey], several test cases are failing. I think for now we can temporarily use isReturnRawValue ? row[rowKey] : \${row[rowKey]}`` and gradually fix the test case issues later.

> 改为 `row[rowKey]` 之后执行测试用例，会有多个测试用例报错。我觉得目前可以暂时先使用 `isReturnRawValue ? row[rowKey] : `${row[rowKey]}`` ，后面再慢慢的修改测试用例报错的问题

SpanManX committed:
fix(components): [table] fix the bug of the expand-row-keys property

Fix the issue where the expand-row-keys property does not support the number type

SpanManX commented on packages/components/table/src/util.ts:
```diff
@@ -177,12 +177,13 @@ export const getColumnByCell = function <T>(
   if (isString(rowKey)) {
     if (!rowKey.includes('.')) {
-      return `${row[rowKey]}`
+      return isReturnRawValue ? row[rowKey] : `${row[rowKey]}`
```
Fixed the multi-level issue and added test cases for multi-level.

> 修改了 multi-level 的问题，并添加了 multi-level 的测试用例

btea reviewed (approved):
Thanks!
